src/CPULSSolver.h

```diff
@@ -76,13 +76,192 @@ class CPULSSolver : public CPUSolver {
+ * @param polar_index polar angle index for this 3D Track
+ * @param track_flux a pointer to the Track's angular flux
+ * @param direction the segment's direction
+ */
+inline void tallyLSScalarFlux(segment* curr_segment, int azim_index,
+                                    int polar_index,
+                                    float* track_flux,
+                                    FP_PRECISION direction[3]) {
+
+  long fsr_id = curr_segment->_region_id;
+  FP_PRECISION length = curr_segment->_length;
+  FP_PRECISION* sigma_t = curr_segment->_material->getSigmaT();
+  FP_PRECISION* position = curr_segment->_starting_position;
+  ExpEvaluator* exp_evaluator = _exp_evaluators[azim_index][polar_index];
+
+  long start_scalar_idx = fsr_id * _num_groups;
+  long start_linear_idx = start_scalar_idx * 3;
+
+  /* Allocate a temporary flux buffer on the stack (free) and initialize it */
+  FP_PRECISION fsr_flux[_num_groups] = {0.0};
+  FP_PRECISION fsr_flux_x[_num_groups] = {0.0};
+  FP_PRECISION fsr_flux_y[_num_groups] = {0.0};
+  FP_PRECISION fsr_flux_z[_num_groups] = {0.0};
+
+  if (_solve_3D) {
+
+    /* Compute the segment midpoint */
+    FP_PRECISION center_x2[3];
+    for (int i=0; i<3; i++)
+      center_x2[i] = 2 * (position[i] + 0.5 * length * direction[i]);
+
+    FP_PRECISION wgt = _quad->getWeightInline(azim_index, polar_index);
+    FP_PRECISION length_2D = exp_evaluator->convertDistance3Dto2D(length);
+
+    // Compute the exponential terms
+    FP_PRECISION exp_F1[_num_groups];
+    FP_PRECISION exp_F2[_num_groups];
+    FP_PRECISION exp_H[_num_groups];
+#pragma omp simd
+    for (int e=0; e < _num_groups; e++) {
+      FP_PRECISION tau = sigma_t[e] * length_2D;
+      exp_evaluator->retrieveExponentialComponents(tau, 0, &exp_F1[e],
+                                                   &exp_F2[e],
+                                                   &exp_H[e]);
+    }
+
+    // Compute the sources
+    FP_PRECISION src_flat[_num_groups];
+    FP_PRECISION src_linear[_num_groups];
+#pragma omp simd
+    for (int e=0; e < _num_groups; e++) {
+      src_flat[e] = _reduced_sources[start_scalar_idx+e];
+      for (int i=0; i<3; i++)
+        src_flat[e] += _reduced_sources_xyz[start_linear_idx+3*e+i] *
+                       center_x2[i];
+      src_linear[e] = _reduced_sources_xyz[start_linear_idx+3*e] *
+                         direction[0];
+      src_linear[e] += _reduced_sources_xyz[start_linear_idx+3*e+1] *
+                         direction[1];
+      src_linear[e] += _reduced_sources_xyz[start_linear_idx+3*e+2] *
+                         direction[2];
+    }
+
+    // Compute the flux attenuation and tally contribution
+#pragma omp simd
+    for (int e=0; e < _num_groups; e++) {
+
+      FP_PRECISION tau = sigma_t[e] * length_2D;
+
+      // Compute the change in flux across the segment
+      exp_H[e] *= length * track_flux[e] * tau * wgt;
+      FP_PRECISION delta_psi = (tau * track_flux[e] - length_2D * src_flat[e]) *
```

Comment:
without delta_psi[e], also vectorized?

Reply:
Turns out yes. 
The key step was to move the track_flux -= delta_psi right after d_psi is calculated instead of after